tests/analysis/type_analysis_test.cpp

```diff
@@ -0,0 +1,240 @@
+    builder.add_container("N", ptr);
+
+    auto& sdfg = builder.subject();
+    auto& root = sdfg.root();
+
+    auto& block = builder.add_block(root);
+    auto& access_node = builder.add_access(block, "i");
+    auto& access_node2 = builder.add_access(block, "N");
+
+    auto& memlet = builder.add_reference_memlet(block, access_node, access_node2, {symbolic::one()}, array2dType);
+
+    analysis::AnalysisManager manager(sdfg);
+    auto& type_analysis = manager.get<analysis::TypeAnalysis>();
+
+    EXPECT_TRUE(type_analysis.get_outer_type("i") != nullptr);
+    EXPECT_TRUE(*type_analysis.get_outer_type("i") == array2dType);
+
+    EXPECT_TRUE(type_analysis.get_outer_type("N") != nullptr);
+    EXPECT_TRUE(*type_analysis.get_outer_type("N") == array2dType);
+}
+
+TEST(TypeAnalysisTest, TaskletReadWrite) {
+    builder::StructuredSDFGBuilder builder("sdfg_1", FunctionType_CPU);
+
+    types::Pointer ptr;
+    types::Scalar base_desc(types::PrimitiveType::Int32);
+    types::Pointer array1dType(base_desc);
+    types::IType& wrapper = array1dType;
+
+    builder.add_container("i", ptr);
+    builder.add_container("N", base_desc);
+
+    auto& sdfg = builder.subject();
+    auto& root = sdfg.root();
+
+    auto& block = builder.add_block(root);
+    auto& access_node = builder.add_access(block, "i");
+    auto& access_node2 = builder.add_access(block, "N");
+
+    auto& tasklet = builder.add_tasklet(block, data_flow::TaskletCode::assign, "_out", {"_in"});
+
+    auto& read_memlet =
+        builder.add_computational_memlet(block, access_node, tasklet, "_in", {symbolic::zero()}, array1dType);
+    auto& write_memlet = builder.add_computational_memlet(block, tasklet, "_out", access_node2, {}, base_desc);
+
+    analysis::AnalysisManager manager(sdfg);
+    auto& type_analysis = manager.get<analysis::TypeAnalysis>();
+
+    EXPECT_TRUE(type_analysis.get_outer_type("i") != nullptr);
+    EXPECT_TRUE(*type_analysis.get_outer_type("i") == array1dType);
+
+    EXPECT_TRUE(type_analysis.get_outer_type("N") != nullptr);
+    EXPECT_TRUE(*type_analysis.get_outer_type("N") == base_desc);
+}
+
+TEST(TypeAnalysisTest, Dereference_Chain) {
+    builder::StructuredSDFGBuilder builder("sdfg_1", FunctionType_CPU);
+
+    types::Pointer ptr;
+    types::Scalar base_desc(types::PrimitiveType::Int32);
+    types::Pointer array1dType(base_desc);
+    types::Pointer array2dType(static_cast<const types::IType&>(array1dType));
+
+    builder.add_container("A", ptr);
+    builder.add_container("B", ptr);
+    builder.add_container("C", ptr);
+
+    auto& sdfg = builder.subject();
+    auto& root = sdfg.root();
+
+    auto& block = builder.add_block(root);
+    auto& access_node = builder.add_access(block, "A");
+    auto& access_node2 = builder.add_access(block, "B");
+
+    auto& memlet = builder.add_dereference_memlet(block, access_node, access_node2, true, array2dType);
+
+    auto& tasklet = builder.add_tasklet(block, data_flow::TaskletCode::assign, "_out", {"0.0"});
```

Comment:
This looks like a race condition. access_node2 and access_node3 use B but are not connected. Either re-use access_node2, which I recommend or add a second block